Lib/test/test_builtin.py

```diff
@@ -328,19 +328,22 @@ def test_compile(self):
 
         codestr = '''def f():
         """doc"""
+        debug_enabled = False
+        if __debug__:
```

Comment:
Why not just
```
debug_enabled = __debug__
```
?

Reply:
... because the tests don't pass that way :)

The __debug__ case gets optimized away for levels 1 & 2, so they're not logically equivalent.

Reply:
worth noting that in a code comment. :)

Reply:
Sorry, I don't see a difference. For level 0 the current code is equivalent to:
```
debug_enabled = False
debug_enabled = True
```
For levels 1 and 2 is is equivalent to:
```
debug_enabled = False
```
How this is different from just`debug_enabled = __debig__`?

Reply:
The optimization levels don't control what `__debug__` evaluates to.

For example, note that the following modified test passes (I added the `__debug__` value along side the `debug_enabled` value in a tuple).

  - `__debug__` evaluates to `True` for cases 0, 1, and 2
  - `debug_enabled` stays`False` for optimization levels 1 & 2 because the `if __debug__` is optimized away

```
$ git diff
diff --git a/Lib/test/test_builtin.py b/Lib/test/test_builtin.py
index 87dcda7b43..32c9c9b044 100644
--- a/Lib/test/test_builtin.py
+++ b/Lib/test/test_builtin.py
@@ -328,9 +328,9 @@ class BuiltinTest(unittest.TestCase):
 
         codestr = '''def f():
         """doc"""
-        debug_enabled = False
+        debug_enabled = (False, __debug__)
         if __debug__:
-            debug_enabled = True
+            debug_enabled = (True, __debug__)
         try:
             assert False
         except AssertionError:
@@ -339,10 +339,10 @@ class BuiltinTest(unittest.TestCase):
             return (False, f.__doc__, debug_enabled)
         '''
         def f(): """doc"""
-        values = [(-1, __debug__, f.__doc__, __debug__),
-                  (0, True, 'doc', True),
-                  (1, False, 'doc', False),
-                  (2, False, None, False)]
+        values = [(-1, __debug__, f.__doc__, (__debug__, __debug__)),
+                  (0, True, 'doc', (True, True)),
+                  (1, False, 'doc', (False, True)),
+                  (2, False, None, (False, True))]
         for optval, assertval, docstring, debugval in values:
             # test both direct compilation and compilation via AST
             codeobjs = []
```

Reply:
Right, you have to remember that `if __debug__:`, as a *phrase*, is treated specially by the *compiler* depending on the optimization level, by dropping the entire suite when the optimization level is higher than 0.  It is `assert` on steroids :)

As Diana said, what value `__debug__` has during runtime is an entirely separate issue.

Reply:
Ah, this is a known bug (see [bpo-22091](https://bugs.python.org/issue22091) and [bpo-27169](https://bugs.python.org/issue27169)).